privacyidea/lib/tokens/foureyestoken.py

```diff
@@ -277,3 +330,181 @@ def authenticate(self, passw, user=None, options=None):
+
+        :param user: the requesting user
+        :type user: User object
+        :param passw: the password: PIN + OTP
+        :type passw: string
+        :param options: additional arguments from the request, which could
+                        be token specific. Usually "transaction_id"
+        :type options: dict
+        :return: return 1 if the answer to the question is correct, -1 otherwise.
+        :rtype: int
+        """
+        options = options or {}
+        r_success = -1
+
+        # fetch the transaction_id
+        transaction_id = options.get('transaction_id')
+        if transaction_id is None:
+            transaction_id = options.get('state')
+
+        # get the challenges for this transaction ID
+        if transaction_id is not None:
+            challengeobject_list = get_challenges(serial=self.token.serial,
+                                                  transaction_id=transaction_id)
+
+            for challengeobject in challengeobject_list:
+                if challengeobject.is_valid():
+                    # challenge is still valid
+                    used_tokens = pickle.loads(challengeobject_list[0].data or pickle.dumps({}))
+                    remaining_realms = self._get_remaining_realms(used_tokens)
+                    r_success = self._authenticate_remaining_realms(passw, remaining_realms, used_tokens, options)
+
+                    if r_success:
+                        challengeobject.set_otp_status(True)
+                    if not r_success:
+                        # increase the received_count
+                        challengeobject.set_otp_status()
+
+        self.challenge_janitor()
+        return r_success
+
+    def create_challenge(self, transactionid=None, options=None):
+        """
+        This method creates a challenge, which is submitted to the user.
+        The submitted challenge will be preserved in the challenge
+        database.
+
+        The challenge is a randomly selected question of the available
+        questions for this token.
+
+        If no transaction id is given, the system will create a transaction
+        id and return it, so that the response can refer to this transaction.
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :type options: dict
+        :return: tuple of (bool, message, transactionid, attributes)
+        :rtype: tuple
+
+        The return tuple builds up like this:
+        ``bool`` if submit was successful;
+        ``message`` which is displayed in the JSON response;
+        additional ``attributes``, which are displayed in the JSON response.
+        """
+        options = options or {}
+        attributes = {}
+        if type(options.get("data")) == dict:
+            # In the special first chal-resp case we do not have pickled data, yet. So we need to convert
+            options["data"] = pickle.dumps(options.get("data"))
```

Comment:
In Python 3 ``pickle.dumps()`` returns a byte object. To put it as a string into the db we need to hexlify it before.

Reply:
You mean byte-string, right?
I am a bit hesitant about hexlifying. The "data" column in the challenge table is 512 characters long.

If we e.g. have such remaining realms to pickle:

~~~~python
s = {"realm1": 1, "someexample.com": 7, "someotherrealm": 1}
~~~~

This is only 70 chars. Hexlifying increases this by 100%. base64 only by 37%.
I.e. with base64 we have 373 available chars, whith hexlify only 256.

Will check this.

Reply:
I think I will go with s.th. like this:

~~~~python
from privacyidea.lib.utils import b64encode_and_unicode
from base64 import b64decode

def pickle_in(s):
    return b64encode_and_unicode(pickle.dumps(s))

def pickle_out(s):
    return pickle.loads(b64decode(s))
~~~~

Reply:
Alternatively to keep the challenge readable in the DB we can use ``json.dumps()``/``json.loads()`` since the object here is already a dictionary.

Reply:
The 4eyes token actually stores the already used token serial numbers in the "data" column like

    '{"realm1": ["pwserial1"]}'

So this sounds like a sensible idea to me.

Reply:
done in ea7f6774